# Connecting to the database

In [1]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

# Data overview

In [7]:
%%sql
SELECT
    name,
    type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

The Chinook record store has just signed a deal with a new record label, and you need to select the first three albums to add to the store, from a list of four. All four albums are by artists who don't have any tracks in the store right now — we have the artists' names and the genre of music they produce:

Artist Name / Genre

Regal / Hip-Hop

Red Tone / Punk

Meteor and the Girls / Pop

Slim Jim Bites / Blues

Write a query that returns each genre with the number of tracks sold in the USA:
- in absolute numbers
- in percentages

Write a paragraph that interprets the data and makes a recommendation for the three artists whose albums we should purchase for the store, based on sales of tracks from their genres.

In [18]:
#We need to JOIN all the information we need from the various tables.

In [15]:
%%sql

WITH usa_sales AS
    (
    SELECT il.* FROM invoice_line il
    INNER JOIN invoice i ON il.invoice_id = i.invoice_id
    INNER JOIN customer c ON i.customer_id = c.customer_id
    WHERE c.country = "USA"
    )
SELECT *
FROM usa_sales us
INNER JOIN track t on t.track_id = us.track_id
INNER JOIN genre g on g.genre_id = t.genre_id;

Done.


[(1, 1, 1158, 0.99, 1, 1158, 'Right Next Door to Hell', 91, 2, 1, None, 182321, 3175950, 0.99, 1, 'Rock'),
 (2, 1, 1159, 0.99, 1, 1159, "Dust N' Bones", 91, 2, 1, None, 298374, 5053742, 0.99, 1, 'Rock'),
 (3, 1, 1160, 0.99, 1, 1160, 'Live and Let Die', 91, 2, 1, None, 184016, 3203390, 0.99, 1, 'Rock'),
 (4, 1, 1161, 0.99, 1, 1161, "Don't Cry (Original)", 91, 2, 1, None, 284744, 4833259, 0.99, 1, 'Rock'),
 (5, 1, 1162, 0.99, 1, 1162, 'Perfect Crime', 91, 2, 1, None, 143637, 2550030, 0.99, 1, 'Rock'),
 (6, 1, 1163, 0.99, 1, 1163, "You Ain't the First", 91, 2, 1, None, 156268, 2754414, 0.99, 1, 'Rock'),
 (7, 1, 1164, 0.99, 1, 1164, 'Bad Obsession', 91, 2, 1, None, 328282, 5537678, 0.99, 1, 'Rock'),
 (8, 1, 1165, 0.99, 1, 1165, 'Back off Bitch', 91, 2, 1, None, 303436, 5135662, 0.99, 1, 'Rock'),
 (9, 1, 1166, 0.99, 1, 1166, "Double Talkin' Jive", 91, 2, 1, None, 203637, 3520862, 0.99, 1, 'Rock'),
 (10, 1, 1167, 0.99, 1, 1167, 'November Rain', 91, 2, 1, None, 537540, 8923566, 0.99, 1, 'Rock'),
 (11, 1, 1168, 0.99, 1, 1168, 'The Garden', 91, 2, 1, None, 322175, 5438862, 0.99, 1, 'Rock'),
 (12, 1, 1169, 0.99, 1, 1169, 'Garden of Eden', 91, 2, 1, None, 161539, 2839694, 0.99, 1, 'Rock'),
 (13, 1, 1170, 0.99, 1, 1170, "Don't Damn Me", 91, 2, 1, None, 318901, 5385886, 0.99, 1, 'Rock'),
 (14, 1, 1171, 0.99, 1, 1171, 'Bad Apples', 91, 2, 1, None, 268351, 4567966, 0.99, 1, 'Rock'),
 (15, 1, 1172, 0.99, 1, 1172, 'Dead Horse', 91, 2, 1, None, 257600, 4394014, 0.99, 1, 'Rock'),
 (16, 1, 1173, 0.99, 1, 1173, 'Coma', 91, 2, 1, None, 616511, 10201342, 0.99, 1, 'Rock'),
 (29, 4, 3448, 0.99, 1, 3448, 'Lamentations of Jeremiah, First Set \\ Incipit Lamentatio', 314, 2, 24, 'Thomas Tallis', 69194, 1208080, 0.99, 24, 'Classical'),
 (30, 4, 2560, 0.99, 1, 2560, 'Violent Pornography', 207, 1, 3, 'Dolmayan, John/Malakian, Daron/Odadjian, Shavo', 211435, 6985960, 0.99, 3, 'Metal'),
 (31, 4, 3336, 0.99, 1, 3336, 'War Pigs', 260, 4, 23, None, 234013, 8052374, 0.99, 23, 'Alternative'),
 (32, 4, 829, 0.99, 1, 829, "Let's Get Rocked", 67, 1, 1, None, 296019, 9724150, 0.99, 1, 'Rock'),
 (33, 4, 1872, 0.99, 1, 1872, 'Attitude', 153, 1, 3, 'Hetfield, Ulrich', 315898, 10335734, 0.99, 3, 'Metal'),
 (34, 4, 748, 0.99, 1, 748, 'Dealer', 58, 1, 1, 'Bolin/Coverdale', 230922, 7591066, 0.99, 1, 'Rock'),
 (35, 4, 1778, 0.99, 1, 1778, "You're What's Happening (In The World Today)", 146, 1, 14, 'Allen Story/George Gordy/Robert Gordy', 142027, 4631104, 0.99, 14, 'R&B/Soul'),
 (36, 4, 2514, 0.99, 1, 2514, 'Spoonman', 203, 1, 1, 'Chris Cornell', 248476, 8289906, 0.99, 1, 'Rock'),
 (37, 5, 1986, 0.99, 1, 1986, 'Intro', 163, 1, 1, 'Kurt Cobain', 52218, 1688527, 0.99, 1, 'Rock'),
 (38, 5, 1987, 0.99, 1, 1987, 'School', 163, 1, 1, 'Kurt Cobain', 160235, 5234885, 0.99, 1, 'Rock'),
 (39, 5, 1988, 0.99, 1, 1988, 'Drain You', 163, 1, 1, 'Kurt Cobain', 215196, 7013175, 0.99, 1, 'Rock'),
 (40, 5, 1989, 0.99, 1, 1989, 'Aneurysm', 163, 1, 1, 'Nirvana', 271516, 8862545, 0.99, 1, 'Rock'),
 (41, 5, 1990, 0.99, 1, 1990, 'Smells Like Teen Spirit', 163, 1, 1, 'Nirvana', 287190, 9425215, 0.99, 1, 'Rock'),
 (42, 5, 1991, 0.99, 1, 1991, 'Been A Son', 163, 1, 1, 'Kurt Cobain', 127555, 4170369, 0.99, 1, 'Rock'),
 (43, 5, 1992, 0.99, 1, 1992, 'Lithium', 163, 1, 1, 'Kurt Cobain', 250017, 8148800, 0.99, 1, 'Rock'),
 (44, 5, 1993, 0.99, 1, 1993, 'Sliver', 163, 1, 1, 'Kurt Cobain', 116218, 3784567, 0.99, 1, 'Rock'),
 (45, 5, 1994, 0.99, 1, 1994, 'Spank Thru', 163, 1, 1, 'Kurt Cobain', 190354, 6186487, 0.99, 1, 'Rock'),
 (46, 5, 1995, 0.99, 1, 1995, 'Scentless Apprentice', 163, 1, 1, 'Nirvana', 211200, 6898177, 0.99, 1, 'Rock'),
 (47, 5, 1996, 0.99, 1, 1996, 'Heart-Shaped Box', 163, 1, 1, 'Kurt Cobain', 281887, 9210982, 0.99, 1, 'Rock'),
 (48, 5, 1997, 0.99, 1, 1997, 'Milk It', 163, 1, 1, 'Kurt Cobain', 225724, 7406945, 0.99, 1, 'Rock'),
 (49, 5, 1998, 0.99, 1, 1998, 'Negative Creep', 163, 1, 1, 'Kurt Cobain', 163761, 5354854, 0.99, 1, 'Rock'),
 (50, 5, 1999, 0.99, 1, 1999, 'Polly', 163, 1, 1, 'Kurt Cobai

In [20]:
#Now that we have all the information we need from the tables, we can make the entire query in only one step.

In [13]:
%%sql

WITH usa_sales AS
    (
    SELECT il.* FROM invoice_line il
    INNER JOIN invoice i ON il.invoice_id = i.invoice_id
    INNER JOIN customer c ON i.customer_id = c.customer_id
    WHERE c.country = "USA"
    )
SELECT
    g.name genre,
    COUNT(us.invoice_line_id) tracks_sold,
    CAST(COUNT(us.invoice_line_id) AS FLOAT) / (
        SELECT COUNT(*) from usa_sales) percentage_sold
FROM usa_sales us
INNER JOIN track t on t.track_id = us.track_id
INNER JOIN genre g on g.genre_id = t.genre_id
GROUP BY 1
ORDER BY 2 DESC
LIMIT 10;

Done.


[('Rock', 561, 0.5337773549000951),
 ('Alternative & Punk', 130, 0.12369172216936251),
 ('Metal', 124, 0.11798287345385347),
 ('R&B/Soul', 53, 0.05042816365366318),
 ('Blues', 36, 0.03425309229305423),
 ('Alternative', 35, 0.03330161750713606),
 ('Latin', 22, 0.02093244529019981),
 ('Pop', 22, 0.02093244529019981),
 ('Hip Hop/Rap', 20, 0.019029495718363463),
 ('Jazz', 14, 0.013320647002854425)]

Based on the sales, the 3 albums we should recommend are: 

Red Tone / Punk (12%)

Slim Jim Bites / Blues (3%)

Meteor and the Girls / Pop (2%)

There is a clear difference between 'Rock' and any other genre, with 53% and trending, current and upcoming rock albums should be looked at with major importance. 